### PR TITLE
Feature/preview service implementation

### PR DESCRIPTION
**Pull request by gsps1**
This is initial check-in for Preview Implementation 
JIRA: https://issues.cask.co/browse/CDAP-5781

This PR covers the following 
- PreviewMain with local modules and few instances shared with Standalone. 
- PreviewMain starts the required services for Preview
- PreviewService starts PreviewHttpHandler which handles requests to start preview, get status of preview, etc
- PreviewManager interface, currently only implementing deploy and running a data pipeline app.
- PreviewDatasetFramework delegating the SYSTEM dataset calls to local (preview space) dsFramework and User namespace calls to actual dsFramework (standalone data space)
- Along with data directory we will have a preview directory in cdap-home when preview is executed. 

Upcoming PR:
- PreviewStore to store intermediate data between stages
- PreviewEmitter and Application changes to emit intermediate data
- PreviewManager completion 
- Test cases

### CODE REVIEW

**Comment by gsps1** [comment]
Build running : http://builds.cask.co/browse/CDAP-DUT4264-1

File: .gitignore [hunk 1]
```diff
@@ -33,6 +33,7 @@ logs/
 /templates/
 /artifacts/
 /data/
+/preview/
```
Comment (chtyim) [inline .gitignore]: Why?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * Preview server modules.
+ */
+public class PreviewServerModule {
+
+  public Module getPreviewModules() {
+    return Modules.combine(new PreviewServiceModule(),
+                           new ConfigStoreModule().getStandaloneModule());
+  }
+
+  private static final class PreviewServiceModule extends AbstractModule {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Extends from `PrivateModule` and only exposes bindings that needed to be exposed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  /**
+   * Configures the Preview pre-start.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    LoggingContextAccessor.setLoggingContext(new ServiceLoggingContext(Id.Namespace.SYSTEM.getId(),
+                                                                       Constants.Logging.COMPONENT_NAME,
+                                                                       Constants.Service.PREVIEW_HTTP));
+    Futures.allAsList(
+      ImmutableList.of(
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: No need to wrap with `ImmutableList`.

File: .gitignore (hunk 1 above)
Comment (gsps1) [inline .gitignore]: fixed. initially had preview dir at same level as data. now moved it inside data dir.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    this.previewManager = previewManager;
+  }
+
+  @POST
+  @Path("/preview")
+  public void startPreview(HttpRequest request, HttpResponder responder,
+                           @PathParam("namespace-id") String namespaceId) throws Exception {
+    String config = request.getContent().toString(Charsets.UTF_8);
+    responder.sendString(HttpResponseStatus.OK,
+                         GSON.toJson(previewManager.start(new NamespaceId(namespaceId), config)));
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: use responder.sendJson. You can pass in the GSON object to be used for serialization.
Benefit of this is that the response type (json) will be set, instead of text/plain (or something like that).

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/PreviewId.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(super.hashCode(), namespace, preview);
+  }
+
+  @Override
+  public Id toId() {
+    return null;
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/PreviewId.java]: throw UnsupportedOperationException? better to fail fast if someone calls this

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java (hunk 4 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: GSON toJson doesn't require type to be given, but the sendJson api needs type if i pass gson and object. so i am less inclined to use that, as it will be complicated(verbose) to provide map type using type-token, etc.

**Comment by gsps1** [comment]
@chtyim addressed your comments, can you review again?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java (hunk 4 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: ok, it may not be a big deal, but it'll probably respond as
text/plain, instead of content type as json, even though it is indeed json

Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: It is important to send the right content type in the response, since content type will be used to parse the response. I think you should be able to use just the class of the object being encoded into JSON as the type in `void sendJson(HttpResponseStatus status, Object object, Type type, Gson gson)`, instead of TypeToken.

Also, can you file an issue for netty-http to make the APIs uniform?

Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: i mean if the type you represent is an interface like map, then you have to use TypeToken to provide the type. but its okay. can be added if its required.

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/PreviewHttpHandler.java:62]: @gsps1 a simple way to avoid TypeToken is to use `object.getClass()`, where `object` is the object being passed in as the second parameter.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      bind(Store.class).to(DefaultStore.class);
+      expose(Store.class);
+
+      bind(ProgramLifecycleService.class).in(Scopes.SINGLETON);
+
+      bind(PreviewServer.class).in(Scopes.SINGLETON);
+      expose(PreviewServer.class);
+
+      bind(PreviewManager.class).to(DefaultPreviewManager.class).in(Scopes.SINGLETON);
+      expose(PreviewManager.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Why need to be exposed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+      Multibinder<HttpHandler> handlerBinder = Multibinder.newSetBinder(
+        binder(), HttpHandler.class, Names.named(Constants.Preview.HANDLERS_BINDING));
+
+      CommonHandlers.add(handlerBinder);
+      handlerBinder.addBinding().to(PreviewHttpHandler.class);
+
+
+      bind(Scheduler.class).to(NoopScheduler.class);
+      expose(Scheduler.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Why need to expose?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Provides
+    @Exposed
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Why exposed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+   * Get the status for the specified {@link PreviewId}.
+   * @param previewId the id of the preview for which status is to be returned
+   * @return the status associated with the preview
+   */
+  PreviewStatus getStatus(PreviewId previewId) throws NotFoundException;
+
+  /**
+   * Get the data associated with the preview.
+   * @param previewId the id associated with the preview
+   * @return the {@link Map} of stage to properties associated with the stage for a given preview
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java]: What do mean by `stage`? Preview doesn't know about stage, isn't it?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.PreviewId;
+import org.apache.twill.api.logging.LogEntry;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Interface used by the Preview REST handler to start the preview and also retrieve the information associated with
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java]: Is it only used by the http handler?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+   * @return the {@link Collection} of metrics emitted during the preview run
+   */
+  Collection<MetricTimeSeries> getMetrics(PreviewId previewId);
+
+  /**
+   * Get the logs for the preview.
+   * @param previewId the id of the preview for which logs to be fetched
+   * @return the logs
+   */
+  List<LogEntry> getLogs(PreviewId previewId);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: For the data type in this method, why some return a `Collection`, some use `List`? I don't see it how they differ from the javadoc.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: I think from the module, only the `PreviewServer` is the only thing that need to be exposed. The StandaloneMain will get an instance of it from the injector and start it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                                                  Constants.AppFabric.DEFAULT_BOSS_THREADS))
+      .setWorkerThreadPoolSize(configuration.getInt(Constants.AppFabric.WORKER_THREADS,
+                                                    Constants.AppFabric.DEFAULT_WORKER_THREADS))
+      .build();
+
+    // Add a listener so that when the service started, register with service discovery.
+    // Remove from service discovery when it is stopped.
+    httpService.addListener(new ServiceListenerAdapter() {
+
+      private List<Cancellable> cancellables = Lists.newArrayList();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: final

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+                                                 Constants.AppFabric.DEFAULT_BACKLOG))
+      .setExecThreadPoolSize(configuration.getInt(Constants.AppFabric.EXEC_THREADS,
+                                                  Constants.AppFabric.DEFAULT_EXEC_THREADS))
+      .setBossThreadPoolSize(configuration.getInt(Constants.AppFabric.BOSS_THREADS,
+                                                  Constants.AppFabric.DEFAULT_BOSS_THREADS))
+      .setWorkerThreadPoolSize(configuration.getInt(Constants.AppFabric.WORKER_THREADS,
+                                                    Constants.AppFabric.DEFAULT_WORKER_THREADS))
+      .build();
+
+    // Add a listener so that when the service started, register with service discovery.
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: Why need to be so complicated? Since this class owns the http service, it can simply start the http server and block on it, then announce it to the discovery service. Similarly for stop as well (cancel service discovery followed by stopping server).

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    if (failureMessage != null ? !failureMessage.equals(that.failureMessage) : that.failureMessage != null) {
+      return false;
+    }
+
+    return status == that.status;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = status.hashCode();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java]: Use `Objects.hashCode`.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java]: from `java.util`

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+      return true;
+    }
+
+    if (!(o instanceof PreviewStatus)) {
+      return false;
+    }
+
+    PreviewStatus that = (PreviewStatus) o;
+
+    if (failureMessage != null ? !failureMessage.equals(that.failureMessage) : that.failureMessage != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java]: Use `Objects.equals`, it handle null for you.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java (hunk 14 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewStatus.java]: Should be `Objects.hash`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  @Override
+  public PreviewId start(NamespaceId namespaceId, String config) throws Exception {
+    final PreviewId previewId = new PreviewId(namespaceId.getNamespace(), RunIds.generate().getId());
+    AppRequest<?> appRequest = GSON.fromJson(config, AppRequest.class);
+    ArtifactSummary artifactSummary = appRequest.getArtifact();
+    Id.Namespace artifactNamespace =
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: For new code, use the `NamespaceId` instead, don't introduce new usages on deprecated classes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewServer.class);
+
+  private final DiscoveryService discoveryService;
+  private final InetAddress hostname;
+  private final ProgramRuntimeService programRuntimeService;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final SystemArtifactLoader systemArtifactLoader;
+
+  private NettyHttpService httpService;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: final. You can construct it in constructor without starting it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+  private Set<HttpHandler> handlers;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: No need if you construct the httpService in constructor.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 8 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: co.cask.cdap.app.guice.InMemoryProgramRunnerModule.DiscoveryServiceAnnouncer announces on "app.bind.address", so we need this to be exposed. let me know if it can be done by another way.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java (hunk 9 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java (hunk 11 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: we want to have getData to use List to make it clear the data is ordered. we use Collection for MetricTimeSeries as ordering doesn't matter there and the underlying Metrics API returns collection.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java (hunk 13 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java]: updated

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 6 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: MapreduceProgramRunner needs Store injected, so exposing store also

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java (hunk 16 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: we call `applicationLifecycleService.deployApp` which still uses old  `Id.Artifact` and so does the ArtifactRepository's `getArtifact` they have to be updated to use the new `ArtifactId`... we should do that in a separate PR.

**Comment by gsps1** [comment]
@chtyim addressed comments, please review again. thanks

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java (hunk 16 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: You can use the new `NamespaceId` in here and call `toId()` when calling those methods.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java (hunk 11 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: Why collection is not order?? Any `Iterable` is ordered as the iteration order.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: To be precise, any immutable iterable is ordered as the iteration order. And I believe this interface returns immutable collection, right?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: And also, based on your comment, it means `getMetrics` may not be ordered (I guess is by time)?

Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewManager.java:75]: we are returning a collection<MetricsTimeSeries>, where MetricsTimeSeries is for a context and metric name, so ordering doesn't matter there. 

Collection can be ordered or unordered, unordered in case of sets. by having list in the API, we are clarifying the properties are ordered. that's the reason behind having it as lists.. If you think its better to make it as Collection, will change it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java (hunk 16 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: can use `NamespaceId` but `ArtifactId` doesn't have `toId()` method, i can make the change for just NamespaceId.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: `ArtifactId` in proto has `toId()` method and that's the replacement for `Id.Artifact`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+  public PreviewStatus getStatus(PreviewId previewId) throws NotFoundException {
+    if (!previewStatusMap.containsKey(previewId)) {
+      throw new NotFoundException(previewId + " not found");
+    }
+    return previewStatusMap.get(previewId);
+  }
+
+  @Override
+  public Map<String, Map<String, List<Object>>> getData(PreviewId previewId) {
+    return null;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java]: Why return null? Even it's to be implemented, it's better to adhere to the contract (non-nullable in this case), hence better return an empty map. Same for the other methods.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+  DefaultPreviewManager(ApplicationLifecycleService applicationLifecycleService,
+                        ProgramLifecycleService programLifecycleService) {
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  @Override
+  public PreviewId start(NamespaceId namespaceId, String config) throws Exception {
+    final PreviewId previewId = new PreviewId(namespaceId.getNamespace(), RunIds.generate().getId());
+    AppRequest<?> appRequest = GSON.fromJson(config, AppRequest.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java:72]: What does it deserialize into?? `AppRequest` is a parameterized class, which has a `T` field inside. Will it deserialize it as a `JsonObject`? If that's the case, use `AppRequest<JsonObject>` instead.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+
+/**
+ * Preview server modules.
+ */
+public class PreviewServerModule {
+
+  public Module getPreviewModules() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Why need this method and an inner class? This class should be just the guice module (meaning whatever you have in the inner class). It's better not to use `Modules.combine`, but instead, when you create the injector, use both modules.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    return Modules.combine(new PreviewServiceModule(),
+                           new ConfigStoreModule().getStandaloneModule());
+  }
+
+  private static final class PreviewServiceModule extends PrivateModule {
+    @Override
+    protected void configure() {
+      bind(PipelineFactory.class).to(SynchronousPipelineFactory.class);
+      expose(PipelineFactory.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Why need expose?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: So why need to expose the `PreviewManager`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 7 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Any comment?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+
+      bind(PreviewManager.class).to(DefaultPreviewManager.class).in(Scopes.SINGLETON);
+      expose(PreviewManager.class);
+
+      Multibinder<HttpHandler> handlerBinder = Multibinder.newSetBinder(
+        binder(), HttpHandler.class, Names.named(Constants.Preview.HANDLERS_BINDING));
+
+      CommonHandlers.add(handlerBinder);
+      handlerBinder.addBinding().to(PreviewHttpHandler.class);
+
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java:72]: Remove extra empty line.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 22 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: this expose has been removed in the latest commit

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 8 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: Move this method to the `ConfigModule` and remove it from any modules that create this binding (this and AppFabricServiceModule at least, not sure if there are other places).

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewApplicationManager.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * This class is concrete implementation of {@link Manager} that deploys an Preview Application.
+ *
+ * @param <I> Input type.
+ * @param <O> Output type.
+ */
+public class PreviewApplicationManager<I, O> implements Manager<I, O> {
+  private final PipelineFactory pipelineFactory;
+  private final NamespacedLocationFactory namespacedLocationFactory;
+  private final CConfiguration configuration;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewApplicationManager.java]: call it `cConf` to standardize.

File: cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+    });
+  }
+
+  /**
+   * This is called by {@link StandaloneMain} with instances that are shared between standalone and preview.
+   * we share DatasetFramework, DiscoveryService (to register preview service), ArtifactRepository.
+   * @param remoteDatasetFramework
+   * @return
+   */
+  public static PreviewMain createPreviewMain(DatasetFramework remoteDatasetFramework,
```
Comment (chtyim) [inline cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java]: Why use a static method? Can't it be just the constructor?

File: cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java [hunk 26]
```diff
@@ -0,0 +1,289 @@
...
+    setConfigurations(cConf, hConf);
+    return new PreviewMain(createPreviewModules(cConf, hConf, remoteDatasetFramework,
+                                                discoveryService, artifactRepository, artifactStore,
+                                                authorizerInstantiator,
+                                                streamAdmin, streamCoordinatorClient, streamConsumerFactory,
+                                                transactionService, transactionManager), cConf);
+  }
+
+  private static void setConfigurations(CConfiguration cConf, Configuration hConf) {
+    // This is needed to use LocalJobRunner with fixes (we have it in app-fabric).
```
Comment (chtyim) [inline cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java]: Is it duplicating the Standalone code?

File: cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java [hunk 27]
```diff
@@ -0,0 +1,289 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.List;
+
+/**
+ * Preview Main.
+ */
+public class PreviewMain {
```
Comment (chtyim) [inline cdap-standalone/src/main/java/co/cask/cdap/PreviewMain.java]: I think it's better not to have this class now, but rather having the Standalone to start the `PreviewServer` (which it will start anything it needed).
This class should be used in future when we run the preview server in distributed mode, which serves as the entry point for that service (and it should probably name as `PreviewServerTwillRunnable`).

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/PreviewId.java [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+   *  Return the preview id
+   */
+  public String getPreview() {
+    return preview;
+  }
+
+  @Override
+  protected Iterable<String> toIdParts() {
+    return null;
```
Comment (chtyim) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/PreviewId.java]: Is the method nullable? And also, shouldn't it return `[namespace, preview]`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java [hunk 29]
```diff
@@ -116,6 +118,61 @@ protected void configure() {
     };
   }
 
+  public Module getPreviewModules(final DatasetFramework remoteDatasetFramework) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java:121]: I think it's better to have preview related stuff under preview packages/classes instead of sitting inside normal classes.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricModules.java [hunk 30]
```diff
@@ -34,6 +37,18 @@ public Module getStandaloneModules() {
     return new DataFabricLocalModule();
   }
 
+
+  public Module getPreviewModules(final TransactionManager transactionManager) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricModules.java:41]: Make preview related stuff group under preview packages/classes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NoopScheduler.java [hunk 31]
```diff
@@ -0,0 +1,119 @@
...
+  @Override
+  public void schedule(Id.Program program, SchedulableProgramType programType, Iterable<Schedule> schedules,
+                       Map<String, String> properties) throws SchedulerException {
+
+  }
+
+  @Override
+  public List<ScheduledRuntime> previousScheduledRuntime(Id.Program program,
+                                                         SchedulableProgramType programType) throws SchedulerException {
+    return null;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NoopScheduler.java]: Does the method allow null? It's better return empty list.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java (hunk 6 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java]: don't have to, i had removed those expose which weren't needed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java (hunk 20 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java:72]: config is deserialized as LinkedTreeMap. cannot use JsonObject. should i use `AppRequest<LinkedTreeMap>`?
